### PR TITLE
Issue-3766 Predicates are not ref counted

### LINKED ISSUE #3766: Dereferencing a predicate doesn't remove it from list of predicates (DEF-2921)
Could give "Could not create more predicates since the buffer is full (64)." errors, even if same predicate is created again.

Repro:

put render.predicate(\{"test"}) in render script's update.

This is kind of an edge case bug.

When it happened:

We had several scenes with different rendering needed for the different scenes. In each scene was a game object that held the reference to the predicates used by that scene. When the scene was loaded the predicates were created. When unloaded the reference was removed together with the game object.

Jumping back and forth between the scenes caused new predicates to be created, resulting in an eventual full buffer.

### PR DESCRIPTION
**Pull request by britzl**
Fixes #3766 

Changed from light user data to user data for the predicates.

### CODE REVIEW

File: engine/render/src/render/render_script.cpp [hunk 1]
```diff
@@ -2213,26 +2248,22 @@ namespace dmRender
...
+
+        luaL_getmetatable(L, RENDER_SCRIPT_PREDICATE);
+        lua_setmetatable(L, -2);
+
+        lua_pushnil(L);  /* first key */
+        while (lua_next(L, 1) != 0)
         {
-            return luaL_error(L, "Could not create more predicates since the buffer is full (%d).", MAX_PREDICATE_COUNT);
+            dmhash_t tag = dmScript::CheckHashOrString(L, -1);
+            AddPredicateTag(*p_predicate, tag);
```
Comment (JCash) [inline engine/render/src/render/render_script.cpp:2262]: Do we want to know if it fails?

**Review by JCash — DISMISSED** [review]
Looks good!
A question regarding hitting the upper limit though.

File: engine/render/src/render/render_script.cpp (hunk 1 above)
Comment (britzl) [reply engine/render/src/render/render_script.cpp:2262]: Yes, maybe. Previously it also silently ignored tags above 64.

Comment (britzl) [reply engine/render/src/render/render_script.cpp:2262]: I thought of returning a result code/boolean and log a warning.

**Review by JCash — APPROVED** [review]
